README
NAME
    App::docsisious - Edit DOCSIS config files on web

VERSION
    0.07

DESCRIPTION
    App::docsisious is a Mojolicious web application for editing DOCSIS
    config files.

    Try out <http://home.thorsen.pm/docsisious> for a demo.

INSTALLATION
      $ curl -L https://cpanmin.us | perl - -M https://cpan.metacpan.org App::docsisious

    You also need SNMP.pm to be able to write "human readable" OIDs. The
    standard behavior is to decode all OIDs as numeric.

      $ sudo apt-get install libsnmp-perl

    If you are using a custom perl build, you have to get SNMP.pm
    <http://www.net-snmp.org/download.html> manually.

SYNOPSIS
      # Step 1:
      $ docsisious --listen http://*:8000
      # Step 2:
      $ open http://localhost:8000

COPYRIGHT AND LICENSE
  Icons
    *   Daniel Bruce <http://www.flaticon.com/free-icon/question-mark_3711>

    *   Dave Gandy
        <http://www.flaticon.com/free-icon/fullscreen-arrows-symbol_25183>

    *   Egor Rumyantsev
        <http://www.flaticon.com/free-icon/settings-work-tool_70367">

    *   Freepik <http://www.flaticon.com/free-icon/save-icon_64052>

    *   Google <http://www.flaticon.com/free-icon/download-button_60721>

  Code
    Copyright (C) 2014, Jan Henning Thorsen

    This program is free software, you can redistribute it and/or modify it
    under the terms of the Artistic License version 2.0.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution App-docsisious

0.11 Not Released
 - Add support for X-Request-Base header

0.10 2017-06-24T09:08:52+0200
 - Fix failing tests on perl 5.26

0.09 2017-06-20T00:05:46+0200
 - Bump required Mojolicious version

0.08 2017-04-29T12:29:54+0200
 - Stop using deprecated "slurp" and "spurt" #3
 - Mojo::Home->parse() is no longer available #3

0.07 2015-11-19T19:28:09+0100
 - App:docsisious::tftpd was not supposed to be released

0.06 2015-11-19T09:26:17+0100
 - Fix version number

0.05 2015-11-19T09:24:29+0100
 - Add optional TFTP capabilities
 - Will also save file as binary on server side
 - Fix 'Use of "do" to call subroutines is deprecated'

0.04 2015-10-15T18:35:57+0200
 - Add link back to main editor page
 - Fix editor on smaller devices

0.03 2015-10-15T10:03:05+0200
 - Change name from App::docsisicious to App::docsisious

0.02 2015-10-15T01:49:52+0200
 - Remove scrollbar in default view
 - Fix executable name
 - Fix SnmpMibObject in example config file

0.01 2015-10-15T01:26:02+0200
 - Add "download" config
 - Add "save" config
 - Add "upload" config



Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use ExtUtils::MakeMaker;
WriteMakefile(
  NAME          => 'App::docsisious',
  AUTHOR        => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE       => 'artistic_2',
  ABSTRACT_FROM => 'lib/App/docsisious.pm',
  VERSION_FROM  => 'lib/App/docsisious.pm',
  EXE_FILES     => [qw(script/docsisious)],
  META_MERGE    => {
    resources => {
      bugtracker => 'https://github.com/jhthorsen/app-docsisious/issues',
      homepage   => 'https://github.com/jhthorsen/app-docsisious',
      repository => 'https://github.com/jhthorsen/app-docsisious.git',
    },
  },
  BUILD_REQUIRES => {'Test::More'         => '0.88'},
  PREREQ_PM      => {'DOCSIS::ConfigFile' => '0.71', 'Mojolicious' => '7.30', 'YAML::XS' => '0.50'},
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);


lib/App/docsisicious.pm
package App::docsisicious;

=head1 NAME

App::docsisicious - Edit DOCSIS config files

=head1 VERSION

0.02

=head1 DESCRIPTION

L<App::docsisicious> is a L<Mojolicious> web application for editing
DOCSIS config files.

=head1 SYNOPSIS

  # Step 1:
  $ docsisicious --listen http://*:8000
  # Step 2:
  $ open http://localhost:8000

=cut

use Mojo::Base -base;
use File::Spec::Functions 'catdir';
use File::Basename 'dirname';
use constant HOME => catdir dirname(__FILE__), 'docsisicious';

our $VERSION = '0.02';

=head1 COPYRIGHT AND LICENSE

=head2 Icons

=over 4

=item * L<Daniel Bruce|http://www.flaticon.com/free-icon/question-mark_3711>

=item * L<Dave Gandy|http://www.flaticon.com/free-icon/fullscreen-arrows-symbol_25183>

=item * L<Egor Rumyantsev|http://www.flaticon.com/free-icon/settings-work-tool_70367">

=item * L<Freepik|http://www.flaticon.com/free-icon/save-icon_64052>

=item * L<Google|http://www.flaticon.com/free-icon/download-button_60721>

=back

=head2 Code

Copyright (C) 2014, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;


lib/App/docsisious/tftpd.pm
package App::docsisious::tftpd;

=head1 NAME

App::docsisious::tftpd - TFTPd server for docsisious

=head1 DESCRIPTION

This class holds logic that can serve config files from L<App::docsisious>.

=head1 SYNOPSIS

It is suggested to use iptables to forward the requests to an unpriviledged
port, so you don't have to run the application as root.

  $ DOCSISIOUS_TFTPD_LISTEN=tftp://*:6969 docsisious daemon

=cut

use Mojo::Base -base;
use File::Spec::Functions qw( catdir catfile );
use Mojo::TFTPd;

=head1 ATTRIBUTES

=head2 listen

  $str = $self->listen;

Returns a listen string. Defaults to C<DOCSISIOUS_TFTPD_LISTEN>.
Example: "tftp://*:12345".

=head2 log

  $log = $self->log;

Holds a L<Mojo::Log> object.

=head2 storage

  $path = $self->storage;

Path to where the files can be found. Default to L<DOCSIS_STORAGE>.

=head2 tftpd

  $tftpd = $self->tftpd;

Holds an instance of L<Mojo::TFTPd>

=cut

has listen  => sub { $ENV{DOCSISIOUS_TFTPD_LISTEN} };
has log     => sub { Mojo::Log->new };
has storage => sub { $ENV{DOCSIS_STORAGE} || ($ENV{HOME} ? catdir $ENV{HOME}, '.docsisious' : '') };
has tftpd => sub { Mojo::TFTPd->new(listen => shift->listen) };

=head1 METHODS

=head2 start

Used to start the TFTPd server.

=cut

sub start {
  my $self   = shift;
  my $listen = $self->listen or return $self;
  my $tftpd  = $self->tftpd;

  Scalar::Util::weaken($self);
  $tftpd->on(error => sub { $self->log->error("[TFTPd] $_[1]") });
  $tftpd->on(
    finish => sub {
      my ($tftpd, $c, $err) = @_;
      if ($err) {
        $self->log->error(sprintf "[TFTPd] %s %s %s FAIL %s", $c->peerhost, $c->type, $c->file, $err);
      }
      else {
        $self->log->info(sprintf "[TFTPd] %s %s %s SUCCESS", $c->peerhost, $c->type, $c->file);
      }
    }
  );

  $tftpd->on(
    rrq => sub {
      my ($tftpd, $c) = @_;
      my $path = catfile $self->storage, $c->file;
      $c->file =~ /^\w+\.bin$/ or die "Invalid file: @{[$c->file]}";
      open my $FH, '<', $path or return;
      $self->log->debug(sprintf "[TFTPd] %s rrq %s", $c->peerhost, $c->file);
      $c->filehandle($FH);
      $c->filesize(-s $path);
    }
  );

  $self->log->info(sprintf "Listening at %s", $self->listen);
  $tftpd->start;

  return $self;
}

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;


lib/App/docsisious.pm
package App::docsisious;
use Mojo::Base -base;

use File::Spec::Functions 'catdir';
use File::Basename 'dirname';
use constant HOME => catdir dirname(__FILE__), 'docsisious';

our $VERSION = '0.07';

1;

=encoding utf8

=head1 NAME

App::docsisious - Edit DOCSIS config files on web

=head1 VERSION

0.07

=head1 DESCRIPTION

L<App::docsisious> is a L<Mojolicious> web application for editing
DOCSIS config files.

Try out L<http://home.thorsen.pm/docsisious> for a demo.

=head1 INSTALLATION

  $ curl -L https://cpanmin.us | perl - -M https://cpan.metacpan.org App::docsisious

You also need L<SNMP.pm|SNMP> to be able to write "human readable" OIDs.
The standard behavior is to decode all OIDs as numeric.

  $ sudo apt-get install libsnmp-perl

If you are using a custom perl build, you have to get L<SNMP.pm|http://www.net-snmp.org/download.html>
manually.

=head1 SYNOPSIS

  # Step 1:
  $ docsisious --listen http://*:8000
  # Step 2:
  $ open http://localhost:8000

=head1 COPYRIGHT AND LICENSE

=head2 Icons

=over 4

=item * L<Daniel Bruce|http://www.flaticon.com/free-icon/question-mark_3711>

=item * L<Dave Gandy|http://www.flaticon.com/free-icon/fullscreen-arrows-symbol_25183>

=item * L<Egor Rumyantsev|http://www.flaticon.com/free-icon/settings-work-tool_70367">

=item * L<Freepik|http://www.flaticon.com/free-icon/save-icon_64052>

=item * L<Google|http://www.flaticon.com/free-icon/download-button_60721>

=back

=head2 Code

Copyright (C) 2014, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut
